discovery_engine_core/ai/ai/src/ranker/system.rs

```diff
@@ -113,14 +113,26 @@ impl Ranker {
     pub(crate) fn log_user_reaction(
         &mut self,
         user_feedback: UserFeedback,
+        title: &str,
         snippet: &str,
         embedding: &Embedding,
         market: &Market,
     ) {
         match user_feedback {
             UserFeedback::Relevant => {
                 let smbert = &self.smbert;
-                let key_phrases = self.kpe.run(snippet).unwrap_or_default();
+                let key_phrases = self
+                    .kpe
+                    .run(snippet)
+                    .or_else(|_| self.kpe.run(format!("{title} {snippet}")))
+                    .map(Into::into)
+                    .unwrap_or_else(|_| {
+                        vec![if !title.is_empty() {
+                            title.to_string()
+                        } else {
+                            snippet.to_string()
+                        }]
```

Comment:
:question: Maybe we need some cleaning of the strings in this case?

Reply:
you mean something like `trim`? `clean_query` is called in another place using this as an input.

Reply:
I was actually thinking about cleaning not trimming as I wasn't sure if having "arbitrary input" as keywords could be a problem at some point in time. But then like you mentioned currently we clean it before using it.